### PR TITLE
Speed up task submissions a bit

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

- port the function descriptor caching from @pcmoritz 
- avoid referencing os.environ when possible
- avoid setting GPU ids when possible

### CODE REVIEW

File: python/ray/remote_function.py [hunk 1]
```diff
@@ -149,8 +151,8 @@ def invocation(args, kwargs):
                     num_return_vals)
             else:
                 object_ids = worker.core_worker.submit_task(
-                    self._function_descriptor.get_function_descriptor_list(),
-                    list_args, num_return_vals, resources)
+                    self._function_descriptor_list, list_args, num_return_vals,
```
Comment (ericl) [inline python/ray/remote_function.py]: @pcmoritz was this the only optimization? I thought there were a couple more.

File: python/ray/function_manager.py [hunk 2]
```diff
@@ -294,6 +294,8 @@ def __init__(self, worker):
         # to a FunctionExecutionInfo object. This should only be used on
         # workers that execute remote functions.
         self._function_execution_info = defaultdict(lambda: {})
+        self._function_descriptors = defaultdict(lambda: {})
```
Comment (pcmoritz) [inline python/ray/function_manager.py]: These two are not needed

**Review by pcmoritz — APPROVED** [review]

File: python/ray/remote_function.py (hunk 1 above)
Comment (pcmoritz) [reply python/ray/remote_function.py]: There was a little more in https://github.com/ray-project/ray/pull/4522 concerning the argument parsing, I'll see if that's still relevant.

**Comment by ericl** [comment]
jenkins retest this please

**Pull request merged**